src/puppeteer_utils.js

```diff
@@ -348,6 +345,58 @@ const compileScript = (scriptString, context = Object.create(null)) => {
     return func;
 };
 
+/**
+ * Extended version of page.goto() allowing to perform requests with HTTP method other than GET,
+ * with custom headers and POST payload. URL, method, headers and payload are taken from
+ * request parameter that must be an instance of Apify.Request class.
+ *
+ * @param {Page} page
+ *   Puppeteer <a href="https://pptr.dev/#?product=Puppeteer&show=api-class-page" target="_blank"><code>Page</code></a> object.
+ * @param {Request} request {@link Request} object.
+ * @param {Object} gotoOptions Custom options for page.goto().
+ * @return {Promise<Response>}
+ *
+ * @memberOf puppeteer
+ */
+export const gotoExtended = async (page, request, gotoOptions = {}) => {
+    checkParamOrThrow(page, 'page', 'Object');
+    checkParamPrototypeOrThrow(request, 'request', Request, 'Apify.Request');
+    checkParamOrThrow(gotoOptions, 'gotoOptions', 'Object');
+
+    const { method, headers, payload } = request;
+
+    if (method !== 'GET' || payload || !_.isEmpty(headers)) {
+        const interceptRequestHandler = async (interceptedRequest) => {
+            const overrides = {};
+
+            if (method !== 'GET') overrides.method = method;
+            if (payload) overrides.postData = payload;
+            if (!_.isEmpty(headers)) overrides.headers = headers;
+
+            await interceptedRequest.continue(overrides);
+            await removeInterceptRequestHandler(page, interceptRequestHandler); // We wan't this to be called only for the initial request.
```

Comment:
I'd also do the URL check and maybe even a [navigation request check](https://pptr.dev/#?product=Puppeteer&version=v1.14.0&show=api-requestisnavigationrequest) and also a check whether its origin is the main frame.

Reply:
Maybe we could just switch lines 376 and 377?

Reply:
How do you get origin (target https://pptr.dev/#?product=Puppeteer&version=v1.14.0&show=api-class-target) from request?

Reply:
Yeah, that makes sense. The removal should be quick enough.